### PR TITLE
Implement configurable keepAlive feature

### PR DESCRIPTION
**Pull request by mhpnguyen**
Implement configurable keepAlive feature.  KeepAlive inside assertion block take precedence over default gateway.properties.

@jsmith2 Can you please review?

### CODE REVIEW

**Comment by mhpnguyen** [comment]
Pull out this PR to wait for future instruction

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/client/interceptor/HeaderRequestOutInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                headers = new HashMap<String, List>();
+            }
+            final Object keepAliveDuration = message.get(NhincConstants.KEEPALIVE_DURATION);
+            LOG.debug("Keep Alive duration {} seconds", keepAliveDuration);
+            if (keepAliveDuration != null) {
+                headers.put(ConnectionType.KEEP_ALIVE.value(),
+                        Collections.singletonList("timeout=" + keepAliveDuration));
+                message.put(Message.PROTOCOL_HEADERS, headers);
+            }
+        } catch (final Exception ce) {
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/client/interceptor/HeaderRequestOutInterceptor.java:76]: Is Exception the actual type that needs to be caught, or can this be replaced with a multicatch?

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        //keepalive from assertion block has precedence over the default properties.
+        // KeepAlive Null means user is not aware of this feature. Will turn on if default indicates so.
+        // KeepAlive < 0 mean user explicitly wants to turn off regardless default value
+        // KeepAlive > 0 mean user explicitly wants to turn on with indicated value.
+        final Integer keepAliveAssertion = assertion.getKeepAlive();
+        final StringBuilder logMsgBuilder = new StringBuilder("KeepAlive statement");
+        logMsgBuilder.append(" for ");
+        logMsgBuilder.append(this.serviceName);
+        logMsgBuilder.append(" service ");
+        if (keepAliveAssertion == null) {// msg comes empty
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java:80]: Stylistic suggestion -- usually, comments should not be on the same line as code.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        bindingProvider.getRequestContext().remove(NhincConstants.KEEPALIVE_DURATION);
+        //keepalive from assertion block has precedence over the default properties.
+        // KeepAlive Null means user is not aware of this feature. Will turn on if default indicates so.
+        // KeepAlive < 0 mean user explicitly wants to turn off regardless default value
+        // KeepAlive > 0 mean user explicitly wants to turn on with indicated value.
+        final Integer keepAliveAssertion = assertion.getKeepAlive();
+        final StringBuilder logMsgBuilder = new StringBuilder("KeepAlive statement");
+        logMsgBuilder.append(" for ");
+        logMsgBuilder.append(this.serviceName);
+        logMsgBuilder.append(" service ");
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java:79]: Stylistic suggestion -- consider chaining appends, e.g.,

final StringBuilder logMsgBuilder = new StringBuilder("KeepAlive statement")
    .append(" for ")
    .append(this.serviceName)
    .append(" service ");

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            if (NullChecker.isNotNullish(sValue)) {
+                defaultKeepAlive = Integer.parseInt(sValue);
+            }
+        } catch (final PropertyAccessException ex) {
+            LOG.warn("Error occurred reading property value from config file ({}): {}", service,
+                    ex.getLocalizedMessage(), ex);
+        } catch (final NumberFormatException nfe) {
+            LOG.warn("Error occurred converting property value from config file ({}): {}", keepAliveKey,
+                    nfe.getLocalizedMessage(), nfe);
+        }
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java:131]: On Exception, consider logging the default value that will be used (or alternatively, a finally statement with a trace output statement containing the defaultKeepAlive value to be returned).

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+/**
+ * This class will allow users to turn on/off KeepAlive feature in http header by detecing keepalive tag inside
+ * assertion block. This also allow users to turn on/off KeepAlive inside gateway.properties. However, the assertion
+ * element will take precedence over the gateway.properties. All values are in seconds. To turn off, either remove
+ * keepAlive elements or using integer < 1
+ *
+ * @author mpnguyen
+ */
+public class KeepAliveServiceEndpointDecorator<T> extends ServiceEndpointDecorator<T> {
+    private static final String KEEP_ALIVE = "keepAliveDuration";
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/messaging/service/decorator/KeepAliveServiceEndpointDecorator.java:50]: Value already exists as a constant in NhincConstants.

File: Product/Production/Common/CONNECTCoreLib/src/test/java/gov/hhs/fha/nhinc/messaging/client/interceptor/HeaderRequestOutInterceptorTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
+/**
+ *
+ */
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/test/java/gov/hhs/fha/nhinc/messaging/client/interceptor/HeaderRequestOutInterceptorTest.java:3]: CONNECT license header should also be applied to test classes.

**Pull request closed without merging**